EntityFrameworkCore.ClickHouse/Query/Internal/ClickHouseQuerySqlGenerator.cs

```diff
@@ -10,145 +8,42 @@ namespace ClickHouse.EntityFrameworkCore.Query.Internal
-                }
-
-                GenerateOrderings(existsExpression.Subquery);
-                GenerateLimitOffset(existsExpression.Subquery);
-
-                if (existsExpression.Subquery.Alias != null)
-                {
-                    subQueryIndent.Dispose();
-
-                    // TODO
-                    //Sql.AppendLine()
-                    //    .Append(")" + AliasSeparator + Sql.DelimitIdentifier(existsExpression.Subquery.Alias));
-                }
+                Sql.Append(" OFFSET ");
+                Visit(selectExpression.Offset);
             }
-
-            return existsExpression;
         }
-        
-        private bool IsNonComposedSetOperation(SelectExpression selectExpression)
-            =>    selectExpression.Offset == null
-               && selectExpression.Limit == null
-               && !selectExpression.IsDistinct
-               && selectExpression.Predicate == null
-               && selectExpression.Having == null
-               && selectExpression.Orderings.Count == 0
-               && selectExpression.GroupBy.Count == 0
-               && selectExpression.Tables.Count == 1
-               && selectExpression.Tables[0] is SetOperationBase setOperation
-               && selectExpression.Projection.Count == setOperation.Source1.Projection.Count
-               && selectExpression.Projection.Select(
-                       (pe, index) => pe.Expression is ColumnExpression column
-                                      && string.Equals(column.Table.Alias, setOperation.Alias, StringComparison.OrdinalIgnoreCase)
-                                      && string.Equals(
-                                          column.Name, setOperation.Source1.Projection[index].Alias, StringComparison.OrdinalIgnoreCase))
-                   .All(e => e);
 
-        private void GenerateList<T>(
-            IReadOnlyList<T> items,
-            Action<T> generationAction,
-            Action<IRelationalCommandBuilder> joinAction = null)
+        protected override Expression VisitSqlParameter(SqlParameterExpression sqlParameterExpression)
         {
-            joinAction ??= (isb => isb.Append(", "));
-
-            for (var i = 0; i < items.Count; i++)
+            if (Sql.Parameters
+                .All(p => p.InvariantName != sqlParameterExpression.Name))
             {
-                if (i > 0)
-                {
-                    joinAction(Sql);
-                }
-
-                generationAction(items[i]);
+                Sql.AddParameter(
+                    sqlParameterExpression.Name,
+                    sqlParameterExpression.Name,
+                    sqlParameterExpression.TypeMapping,
+                    sqlParameterExpression.IsNullable);
             }
+
+            // see https://github.com/DarkWanderer/ClickHouse.Client/wiki/SQL-Parameters
+            Sql.Append("{").Append($"{sqlParameterExpression.Name}:{sqlParameterExpression.TypeMapping.StoreType}").Append("}");
```

Comment:
Can you use here https://github.com/denis-ivanov/EntityFrameworkCore.ClickHouse/blob/master/EntityFrameworkCore.ClickHouse/Storage/Internal/ClickHouseSqlGenerationHelper.cs#L20 ?

Reply:
IRelationalCommandBuilder contain StringBuilder wrapper, not StringBuilder, So I change GenerateParameterNamePlaceholder return string rather than append straightforwardly